primitives/src/calendar.rs

```diff
@@ -0,0 +1,340 @@
+                        let elapsed_months =
+                            year_diff * 12 + month_now_0_indexed - month_start_0_indexed;
+                        let elapsed_periods = elapsed_months / multiplier;
+                        let next_period_months = multiplier * (elapsed_periods + 1);
+                        // The month of the next period counting from the beginning of `year_start`.
+                        let denormalized_next_period_month =
+                            month_start_0_indexed + next_period_months;
+                        let next_period_year =
+                            year_start + i32::try_from(denormalized_next_period_month / 12).ok()?;
+                        // Month in base 12.
+                        let next_period_month = denormalized_next_period_month % 12;
+                        let max_month_days = match next_period_month {
+                            // Handle 30-day months.
+                            3 | 5 | 8 | 10 => 30,
+                            // Handle February.
+                            1 if is_leap_year(next_period_year) => 29,
+                            1 => 28,
+                            // No month has more than 31 days.
+                            _ => 31,
+                        };
+                        let next_period_day = u32::min(day_start, max_month_days);
+                        let date_next = NaiveDate::from_ymd(
+                            i32::try_from(next_period_year).ok()?,
+                            // Convert months back to calendar numbering.
+                            next_period_month + 1,
+                            next_period_day,
+                        );
+                        u64::try_from(date_next.and_time(date_time_start.time()).timestamp()).ok()
+                    }
+                    VariableCalendarUnit::Year => {
+                        // Convert the multiplier to match the type of year.
+                        let multiplier = i32::try_from(multiplier).ok()?;
+                        let elapsed_periods: i32 = (date_now.year() - year_start) / multiplier;
+                        let next_period_year = year_start + multiplier * (elapsed_periods + 1);
+                        let next_period_day = if month_start == 2 && !is_leap_year(next_period_year)
+                        {
+                            // Handle February in non-leap years.
+                            u32::min(day_start, 28)
+                        } else {
+                            day_start
+                        };
+                        let date_next = NaiveDate::from_ymd(
+                            next_period_year,
+                            // Convert months back to calendar numbering.
+                            month_start,
+                            next_period_day,
+                        );
+                        u64::try_from(date_next.and_time(date_time_start.time()).timestamp()).ok()
```

Comment:
Let's extract this out of each match arm, returning the `NaiveDate::from_ymd(...)` from each instead.